File: app/src/main/java/com/jinhanexample/mvvmSample/java/author/ui/AuthorRecyclerViewAdapter.java
package com.jinhanexample.mvvmSample.java.author.ui;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.jinhanexample.R;
import com.jinhanexample.mvvmSample.java.author.data.AuthorEntity;
import com.jinhanexample.mvvmSample.java.author.listener.LongClickListener;

import java.util.ArrayList;
import java.util.List;

public class AuthorRecyclerViewAdapter extends RecyclerView.Adapter<AuthorRecyclerViewAdapter.ViewHolder> {

    private List<AuthorEntity> authorEntityList = new ArrayList<>();
    private LongClickListener longClickListener;

    public AuthorRecyclerViewAdapter(LongClickListener longClickListener) {
        this.longClickListener = longClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.author_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(authorEntityList.get(position));
    }

    @Override
    public int getItemCount() {
        return authorEntityList.size();
    }

    public void addItem(List<AuthorEntity> authorEntityList) {

        this.authorEntityList = authorEntityList;
        notifyDataSetChanged();
    }

    class ViewHolder extends RecyclerView.ViewHolder {


        public ViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        TextView authorName = itemView.findViewById(R.id.authorName);
        TextView authorBooks = itemView.findViewById(R.id.authorBooks);
        TextView authorNation = itemView.findViewById(R.id.authorNation);

        void bind(AuthorEntity authorEntity) {
            authorName.setText(authorEntity.getName());
            authorBooks.setText(authorEntity.getBooks());
            authorNation.setText(authorEntity.getNation());

            itemView.setOnClickListener(view -> {
                longClickListener.onLongClickListener(authorEntity);
            });
        }
    }
}

File: app/src/main/java/com/jinhanexample/mvvmSample/kotlin/author/ui/AuthorRecyclerViewAdapter.kt
package com.jinhanexample.mvvmSample.kotlin.author.ui

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.jinhanexample.R
import com.jinhanexample.mvvmSample.kotlin.author.data.AuthorEntity


//RecyclerViewAdapter에 대한 설명은 생략하겠습니다.
//CompletionBlock을 사용해 Long클릭을 했을 때 delete Dialong를 띄워 데이터를 삭제하겠습니다.
class AuthorRecyclerViewAdapter(val onLongClick : (AuthorEntity) -> Unit) : RecyclerView.Adapter<AuthorRecyclerViewAdapter.ViewHolder>() {
    private var authorEntity: List<AuthorEntity> = listOf()


    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val authorName = itemView.findViewById<TextView>(R.id.authorName)
        val authorBooks = itemView.findViewById<TextView>(R.id.authorBooks)
        val authorNation = itemView.findViewById<TextView>(R.id.authorNation)

        fun bind(authorEntity: AuthorEntity) {
            authorName.text = authorEntity.name
            authorBooks.text = authorEntity.books
            authorNation.text = authorEntity.nation

            //아이템뷰를 클릭했을 때 onLongClick을 이용해 authorEntity데이터 전달
            itemView.setOnClickListener {
                onLongClick(authorEntity)
            }

        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context).inflate(R.layout.author_list_item, parent, false)
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(authorEntity[position])
    }

    override fun getItemCount(): Int {
        return authorEntity.size
    }

    fun addItem(authorEntity: List<AuthorEntity>){
        this.authorEntity = authorEntity
        notifyDataSetChanged()

    }
}